route/bgp_path_manager.go

```diff
@@ -0,0 +1,63 @@
+package route
+
+import (
+	"log"
+	"sync"
+)
+
+// BGPPathManager is a component used to deduplicate BGP Path objects
+type BGPPathManager struct {
+	paths map[BGPPath]*BGPPathCounter
+	mu    sync.Mutex
+}
+
+// BGPPathCounter couples a counter to a particular path
+type BGPPathCounter struct {
+	usageCount uint64
+	path       *BGPPath
+}
+
+// NewBGPPathManager creates a new BGP Path Manager
+func NewBGPPathManager() *BGPPathManager {
+	m := &BGPPathManager{}
+	return m
+}
+
+func (m *BGPPathManager) pathExists(p BGPPath) bool {
+	if _, ok := m.paths[p]; !ok {
+		return false
+	}
+
+	return true
+}
```

Comment:
i woulud rename this function to `lookup` and return `nil` in case the element is not in the map, and a (pointer to the) object if the key exists.
This would speedup the `AddPath` method, since you don't need to lookup the key twice in the map.

Depending on the used hash function this could have a performance impact in case of 700k calls to `AddPath`

Reply:
Done